http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpRequest.java

```diff
@@ -18,40 +18,100 @@
+     *
+     * <p>Important Note: AWS signing DOES NOT include the port when the request is signed if the default port for the protocol is
+     * being used. When sending requests via http over port 80 or via https over port 443, the URI or host header MUST NOT include
+     * the port or a signature error will be raised from the service for signed requests. HTTP plugin implementers are encouraged
+     * to use the {@link #toUri()} method for generating the URI to use for communicating with AWS to ensure the URI used in the
+     * request matches the URI used during signing.</p>
+     *
+     * @return The port that should be used for HTTP communication.
+     */
+    int port();
 
     /**
-     * Returns the path to the resource being requested.
+     * Returns the URL-encoded path that should be used in the HTTP request.
+     *
+     * <p>If a path is configured, the path will always start with '/' and never end with '/' (eg. "/my/path").
+     * If a path is not configured, this will always return empty-string (ie. "").</p>
      *
      * @return The path to the resource being requested.
      */
-    String getResourcePath();
+    String encodedPath();
 
     /**
-     * Returns a map of all parameters in this request.
-     * <br/>
-     * Should never be null, if there are no parameters an empty map is returned.
+     * Returns a map of all non-URL encoded parameters in this request. HTTP plugins can use
+     * {@link SdkHttpUtils#encodeQueryParameters(Map)} to encode parameters into map-form, or
+     * {@link SdkHttpUtils#encodeAndFlattenQueryParameters(Map)} to encode the parameters into uri-formatted string form.
+     *
+     * <p>This will never be null. If there are no parameters an empty map is returned.</p>
      *
-     * @return A map of all parameters in this request.
+     * @return An unmodifiable map of all non-encoded parameters in this request.
      */
-    @ReviewBeforeRelease("Should this be 'query parameters'? It took me a while to figure out what an HTTP parameter was.")
-    Map<String, List<String>> getParameters();
+    Map<String, List<String>> rawQueryParameters();
 
     /**
-     * Returns the service endpoint (ex: "https://ec2.amazonaws.com") to which
-     * this request should be sent.
+     * Convert this HTTP request's protocol, host, port, path and query string into a properly-encoded URI string that matches the
+     * URI string used for AWS request signing.
      *
-     * @return The service endpoint to which this request should be sent.
+     * <p>The URI's port will be missing (-1) when the {@link #port()} is the default port for the {@link #protocol()}. (80 for
+     * http and 443 for https). This is to reflect the fact that request signature does not include the port.</p>
+     *
+     * @return The URI for this request, formatted in the same way the AWS HTTP request signer uses the URI in the signature.
      */
-    URI getEndpoint();
+    default URI toUri() {
```

Comment:
toUri doesn't seem like the right name here. getUri?

Reply:
+1